sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,250 @@
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if estimator is None:
+        raise ValueError("base_estimator cannot be None")
+
+    if not hasattr(estimator, "predict_proba"):
+        raise ValueError("base_estimator should implement predict_proba!")
+
+
+class SelfTrainingClassifier(BaseEstimator):
+    """Self-training classifier
+
+    Parameters
+    ----------
+    base_estimator : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``self.base_estimator_`` attribute.
+
+    threshold : float
+        Threshold above which predictions are added to the labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : integer
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0.
+
+    Attributes
+    ----------
+    base_estimator_ : estimator object
+        The fitted estimator.
+
+    y_labeled_ : array, shape = (n_samples,)
+        The labels used for the final fit of the classifier.
+
+    y_labeled_iter_ : array, shape = (n_samples,)
+        The iteration in which each sample was labeled. When a sample has
+        iteration 0, the sample was already labeled in the given dataset. When
+        a sample has iteration -1, the sample was not labeled in any iteration.
+
+    n_iter_ : int
+        The amount of iterations the classifier during fitting.
+
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTrainingClassifier
+    >>> from sklearn.svm import SVC
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    _estimator_type = "classifier"
```

Comment:
Inheriting from `ClassifierMixin` would avoid you this, but I'm not sure if this is worth it or even wanted. Would need additional input.

Reply:
I originally inherited from `ClassifierMixin`, but removed it since the `score` method it provides would always be present, while `test_metaestimators.py` expects it to be hidden when the base estimator doesn't have a `score` method. So now I delegate score to the base estimator, which should be functionally equivalent.

Of course this is just a minor thing, so if there are other reasons for inheriting from `ClassifierMixin` it's no issue.